### PR TITLE
#4838 api for event comment likes and replies

### LINKED ISSUE #4838: [GreenCity][Event-comment-controller] Add controllers for returning replies for comment, count replies, add likes, count likes

### PR DESCRIPTION
**Pull request by Maksym-Lenets**
## Summary Of Issue :
Need to add endpoints to EventCommentController that can manage replies and likes for such comments

**Issue Link**
https://github.com/ita-social-projects/GreenCity/issues/4838

## Summary Of Changes :
1) added opportunity to add and count replies and likes
2) EventComment updated with one to many relations for replies  
3) updated EventComment's DTO, VO and their mappers 
4) updated database with new "delete" field for event_comments table and added table for storing likes
5) updated deleting approach for comments and their replies (now comments just marked as "deleted")
6) refactored old tests and added new ones

## CHECK LIST
- [x] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ Github Issue number & title in PR title (ISSUE-: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

**Review by sergoliarnik — APPROVED** [review]

**Review by juseti — APPROVED** [review]

**Review by ABbondar — APPROVED** [review]

**Review by LiliaMokhnatska — APPROVED** [review]

File: dao/src/main/java/greencity/repository/EventCommentRepo.java [hunk 1]
```diff
@@ -16,15 +18,44 @@ public interface EventCommentRepo extends JpaRepository<EventComment, Long> {
      */
     @Query(value = "select count(ec.id) from events_comment ec"
         + " join events event on event.id = ec.event_id"
-        + " where event.id = :event", nativeQuery = true)
+        + " where event.id = :event and deleted = false", nativeQuery = true)
     int countEventCommentsByEvent(Event event);
```
Comment (RomaMocherniuk) [inline dao/src/main/java/greencity/repository/EventCommentRepo.java]: please update java doc and method name that we do not return deleted comments

File: service/src/test/java/greencity/mapping/EventCommentDtoMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@ExtendWith(MockitoExtension.class)
+class EventCommentDtoMapperTest {
+    @InjectMocks
+    EventCommentDtoMapper eventCommentDtoMapper;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/mapping/EventCommentDtoMapperTest.java]: `private`

File: service/src/main/java/greencity/service/EventCommentServiceImpl.java [hunk 3]
```diff
@@ -165,41 +198,116 @@ public void update(String commentText, Long id, UserVO userVO) {
...
-        eventCommentRepo.save(EventComment.builder().text(replyText).parentComment(eventParentComment).build()
-            .setUser(modelMapper.map(userVO, User.class))
-            .setEvent(eventParentComment.getEvent()));
+    @Override
+    public AmountCommentLikesDto countLikes(Long commentId, UserVO userVO) {
+        EventComment comment = eventCommentRepo.findByIdAndDeletedFalse(commentId).orElseThrow(
+            () -> new NotFoundException(ErrorMessage.EVENT_COMMENT_NOT_FOUND_BY_ID + commentId));
+
+        UserVO user = userVO == null ? UserVO.builder().build() : userVO;
+        boolean isLiked = comment.getUsersLiked().stream().anyMatch(u -> u.getId().equals(user.getId()));
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/EventCommentServiceImpl.java]: `isLiked = userVO == null ? false : comment.getUsersLiked().stream().anyMatch(u -> u.getId().equals(userVO()));`

File: core/src/main/java/greencity/config/SecurityConfig.java [hunk 4]
```diff
@@ -146,6 +146,10 @@ protected void configure(HttpSecurity http) throws Exception {
                 "/econews/comments/active",
                 "/events/comments/active",
                 "/events/comments/count/{eventId}",
+                "/events/comments/replies/active/{parentCommentId}",
+                "/events/comments/replies/active/count/{parentCommentId}",
+                "/events/comments/like",
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/config/SecurityConfig.java]: How does an unauthorized user like some comments?

File: core/src/test/java/greencity/controller/EventCommentControllerTest.java [hunk 5]
```diff
@@ -132,4 +145,250 @@ void countComments() {
...
+
+        UserVO userVO = getUserVO();
+        when(userService.findByEmail(anyString())).thenReturn(userVO);
+
+        Pageable pageable = PageRequest.of(pageNumber, pageSize);
+        PageableDto<EventCommentDto> eventCommentReplies = getPageableEventCommentDtos();
+
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.findAndRegisterModules();
+        ObjectWriter ow = objectMapper.writer();
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/controller/EventCommentControllerTest.java]: the redundant variable
you can use `objectMapper. writeValueAsString(...)`

File: core/src/test/java/greencity/controller/EventCommentControllerTest.java [hunk 6]
```diff
@@ -132,4 +145,250 @@ void countComments() {
...
+
+    @Test
+    @SneakyThrows
+    void findAllActiveReplies() {
+        Long parentCommentId = 1L;
+        int pageNumber = 0;
+        int pageSize = 20;
+
+        UserVO userVO = getUserVO();
+        when(userService.findByEmail(anyString())).thenReturn(userVO);
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/controller/EventCommentControllerTest.java:156]: please add `verify`

File: service-api/src/main/java/greencity/dto/eventcomment/EventCommentVO.java [hunk 7]
```diff
@@ -32,9 +34,17 @@ public class EventCommentVO {
...
 
     private UserVO user;
 
     private EventVO event;
+
+    private boolean deleted;
+
+    private boolean currentUserLiked = false;
+
+    private Set<UserVO> usersLiked = new HashSet<>();
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/eventcomment/EventCommentVO.java:49]: should it be UserDto?
`UserVO` contains a lot of personal data

File: service-api/src/main/java/greencity/dto/eventcomment/EventCommentVO.java [hunk 8]
```diff
@@ -32,9 +34,17 @@ public class EventCommentVO {
...
+
     private List<EventCommentVO> comments = new ArrayList<>();
 
     private UserVO user;
 
     private EventVO event;
+
+    private boolean deleted;
+
+    private boolean currentUserLiked = false;
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/eventcomment/EventCommentVO.java]: `boolean` default value is `false`

**Review by RomaMocherniuk — APPROVED** [review]

**Pull request merged**